### DISCUSSION TITLE
UI related stuff / questions /etc

### DISCUSSION THREAD

**Discussion by Wiki-the-Tinkerer** [martin2250/OpenCNCPilot#178] [General]
Hi folx, I'm new to OpenCNCPilot so please excuse me if I mention things already discussed/planned/issued etc. Just wanted to dump som first use thoughts, as I have observed that initial thoughts/questions often are valuable in UI design, as you tend to forget the dummy viewpoint, after working on something for long. I have used many GRBL senders, but this one is my new go-to tool, so I just want to contribute something back.

**Comment by Wiki-the-Tinkerer** [comment]
1.5.12.0 - thoughts dump

File menu - What is the relation between the "Pause", and "[]Pause" in the File menu?

File menu vs Main center top menu - What is the relation between the "Start" & "Pause" functions in File compared to the "Hold" and "Start" in the center menu?  (I sometimes click start in files, and then start in main, and then start in files again, and at some point the job starts, but it's not clear what does what

Overrides menu - minor thing, but took me a while to realize that the "Rapid" override did not add speed (i.e. +50 and +25%), but override it only downwards to "absolute" 50 or 25. This as I had used the other overrides that have scale. Maybe add also +25, +50 etc? 

Overrides menu - + to the left, and - to the right. There may be some perfectly good explanation for this, but usually in western devs the direction is opposite - i.e. increases to the right, decreases to the left.    

Log of console - for troubleshooting a job, where something went wrong its often helpful to see the console log from the machine (the one that scrolls by very fast). I'm sure its possible to find it somehow, but it might be very helpful to have a button to open it (responses related to the commands sent)

File menu - "Clear": I have pressed it instinctively a couple times, expecting functionality like "reset" in e.g. Candle - i.e. to "clear the job", and return to start of the code, but it deletes the code. Figured out that "Go to 1" works fine, but maybe a button "Start over", "restart" or similar could be clearer? 

Lost connection - Would be great if there would be a visible (persistant until action) alert if the sender loses connection to the machine in the middle of a job. Happened several times as I was testing a new DLC32 v2 board, that apparently is VERY sensitive to power fluctuations. Currently the status of course stops running, but otherwise its not very clear. what is going on. and clicking around with some expensive stock on the table is a bit.. worrying. 

Viewport movement: I really like the principle of not using the mouse for viewport and especially jogging, where it is much safer. However, took a while before i figured out the viewpoint using keys, as the focus is also needed. Maybe a transparent icon on the work space, with keys? (and mousover hint)

Hope any of this is useful, and pls tell me if this is not the right place for this kinda stuff.
/W

**Comment by martin2250** [comment]
Hi Wiki,

> File menu - What is the relation between the "Pause", and "[]Pause" in the File menu?

check the tooltip, the checkbox controls wether or not it should pause the program on M0/M1/M30.

> File menu vs Main center top menu - What is the relation between the "Start" & "Pause" functions in File compared to the "Hold" and "Start" in the center menu? (I sometimes click start in files, and then start in main, and then start in files again, and at some point the job starts, but it's not clear what does what

The file menu "Pause" will just stop sending gcode to grbl, but everything that is in the grbl buffer will still be executed.
The center menu "Hold" will send a real-time hold command to grbl, so the machine will stop immediately (but you can't start doing other things like jogging)

> Overrides menu - minor thing, but took me a while to realize that the "Rapid" override did not add speed (i.e. +50 and +25%), but override it only downwards to "absolute" 50 or 25. This as I had used the other overrides that have scale. Maybe add also +25, +50 etc?

That's not easily possible, that's the way grbl works. otherwise I'd have done it that way 😉 
In general, I want OCP to be as close to "talking to grbl directly" as it can be and not abstract such things away.

> Overrides menu - + to the left, and - to the right. There may be some perfectly good explanation for this, but usually in western devs the direction is opposite - i.e. increases to the right, decreases to the left.

True, I didn't think about that. But changing that now would probably also confuse the people already using it. I probably won't change that unless some of the "old folks" agree

> Log of console - for troubleshooting a job, where something went wrong its often helpful to see the console log from the machine (the one that scrolls by very fast). I'm sure its possible to find it somehow, but it might be very helpful to have a button to open it (responses related to the commands sent)

You can use the "log traffic to file" option in the settings menu

> File menu - "Clear": I have pressed it instinctively a couple times, expecting functionality like "reset" in e.g. Candle - i.e. to "clear the job", and return to start of the code, but it deletes the code. Figured out that "Go to 1" works fine, but maybe a button "Start over", "restart" or similar could be clearer?

I'll write that down for the next version, this has bugged me a few times as well.

> Lost connection - Would be great if there would be a visible (persistant until action) alert if the sender loses connection to the machine in the middle of a job. Happened several times as I was testing a new DLC32 v2 board, that apparently is VERY sensitive to power fluctuations. Currently the status of course stops running, but otherwise its not very clear. what is going on. and clicking around with some expensive stock on the table is a bit.. worrying.

Good idea. In general I'd say you should make your hardware reliable, but in the end it can always happen. Writing that down as well 👍🏻 

> Viewport movement: I really like the principle of not using the mouse for viewport and especially jogging, where it is much safer. However, took a while before i figured out the viewpoint using keys, as the focus is also needed. Maybe a transparent icon on the work space, with keys? (and mousover hint)

I like the idea of making that more obvious for new users, but I don't quite understand your suggestion, can you elaborate?


Thanks for the input :)

Cheers,
Martin

**Reply by Wiki-the-Tinkerer** [reply]
Thanks for the fast reply, Martin :). Dropping notes a few at a time, when free time

> -File menu - What is the relation between the "Pause", and "[]Pause" in the File menu?
> --check the tooltip, the checkbox controls wether or not it should pause the program on M0/M1/M30.

Ok.. um.. the execution (Cut) stops for M0 in any case, doesnt it (after having started from "Send-"start)? If its only the send that should stop, it also seems to send a few lines past the M0. If the checkbox is unmarked, one center "Cut-"Start click seems to continue the cut, but if checked, it takes a lot of clicks on the center Start, before the cut continues (possibly because of missing F values on singuar moving code lines?? as it executes only one line per "cut-"start press). I get it to work, just wondering.

> The file menu "Pause" will just stop sending gcode to grbl, but everything that is in the grbl buffer will still be executed.The center menu "Hold" will send a real-time hold command to grbl

Right, makes sense, it was actually just me and a bad formulated question. The real question should have been about the two "Start"s. it is a bit unclear, as "Send-"start both starts sending, but also starts executing (cutting), so the "cut-"start seems redundant - but it is not grayed out. 
But if you start cutting with the "Send-"start, and e.g. reach a M0 pause, the cutting stops, and you need to push the "Cut-"start to continue. And after that, the "cut-" start button is still available, whereas the "Send-" start button grays out after being clicked (i.e. its function is active). Also see the "Pause" scenarios above.  

> Overrides menu - rapids
Of curse - didnt realize it was the "real" rapids, that are not usually available. Great that it is, though. Maybe ordering them Feed, Spindle, Rapid would make sense? Would seem the Frequency of use - order?

**Comment by Wiki-the-Tinkerer** [comment]
Was thinking about the overrides, and came to the conclusion that sliders might be the clearest component, showing both range and position. 
And then downloaded the newest version of candle, as changed boards to some machines, so could update at the same time.. and lo and behold, it has also added rapids, and a enhanced slider system that displays both current position and the desired override position (smart). And uses the checkboxes as "reset to 100" without resetting, so its easy to return to the previous setting. 
Dont think its easy to make a better override scaling solution than that, might be worth considering.

**Comment by Wiki-the-Tinkerer** [comment]
Back again.. re: Viewport movement: 
(after a while I even figured out that the right button also turns the view, dummy as i am ;) 

What i thought about was something like the CAM/3d programs have, a turning sphere or cube.. quick mockup below. Visible on mouseover of table (=when active), tool tip like the others when mouseover icon, hidable from view settings (might also make sense (?) to separate view settings from the de-bugging) 
  

![tilting](https://github.com/martin2250/OpenCNCPilot/assets/140900357/ff467a03-4aeb-45cc-a7dc-55ced5ee553a)

**Reply by deHarro** [reply]
Hi @stefan63!

Perhaps a bit late, just triggered by your last post...
I implemented a view cube (just for fun as I was fiddling around with the code already, remembering having read about that somewhere - yes, over here ;)

Have a look at my [fork](https://github.com/deHarro/OpenCNCPilot/tree/master/OpenCNCPilot).

The cube is only displaying the current orientation, cannot be used to drag the view with the mouse.
Clicking on one of the faces turns the view to show this side to the user.
And yes, it is turned on via Debug ;)

**Comment by stefan63** [comment]
Any chance to add stylesheet to improve estetics?

**Reply by deHarro** [reply]
For that, I think you have to wait for Martin... but I guess, that will not happen. OCP does what it is intended for and the graphics is not so bad. I see numerous implementations of g-code streamers all over, coming with really stale GUIs.

By the way... what do you mean, concerning esthetics improvement? 

Separating "view" from "debug": Ok, that is my fault. As stated elswhere, I'm no C# coder and so I was where looking for a non intrusive spot to place my additional gadgets.
"Flicker": Perhaps that is implied by your graphics card?
I use the graphics core in the CPU (Intel and AMD on several computers) and have no flicker at all when moving the vieport of OCP.
Ok, my hardware is non of the weaker end of performance...

Have fun!
